File: python/tests/test_strings.py
# Copyright (c) 2023-2025, NVIDIA CORPORATION. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

import pyarrow as pa
import pytest

from legate_dataframe import LogicalColumn
from legate_dataframe.lib import strings
from legate_dataframe.testing import assert_matches_polars

match_funcs = ["starts_with", "ends_with", "match_substring", "match_substring_regex"]


@pytest.mark.parametrize("match_func", match_funcs)
@pytest.mark.parametrize("pattern", ["", "hel", "o", "o.+d"])  # codespell:ignore
def test_matches(match_func, pattern):
    arr = pa.array(
        ["hello", "olhel", "world", "wo.+d"] * 2,  # codespell:ignore
        mask=[True, True, True, True, False, False, False, False],
    )
    column = LogicalColumn.from_arrow(arr)
    result = strings.match(match_func, column, pattern)
    expected = getattr(pa.compute, match_func)(arr, pattern)
    assert expected == result.to_arrow()


@pytest.mark.parametrize("match_func", match_funcs)
def test_matches_empty(match_func):
    column = LogicalColumn.from_arrow(pa.array([], "string"))
    result = strings.match(match_func, column, "o")
    assert result.to_arrow() == pa.array([], "bool")


def test_bad_function():
    column = LogicalColumn.from_arrow(pa.array(["hello", "world", "help", "hero"]))
    with pytest.raises(ValueError):
        strings.match("bad_function", column, "o")


@pytest.mark.parametrize(
    "pl_func",
    [
        lambda x, p: getattr(x.str, "starts_with")(p),
        lambda x, p: getattr(x.str, "ends_with")(p),
        lambda x, p: getattr(x.str, "contains")(p),
        lambda x, p: getattr(x.str, "contains")(p, literal=True),
    ],
)
@pytest.mark.parametrize("pattern", ["", "hel", "o", "o.+d", "l$"])  # codespell:ignore
def test_polars_matches(pl_func, pattern):
    pl = pytest.importorskip("polars")

    arr = pa.array(
        ["hello", "olhel", "world", "wo.+d"] * 2,  # codespell:ignore
        mask=[True, True, True, True, False, False, False, False],
    )

    q = pl.LazyFrame({"a": arr})
    q = q.with_columns(matches=pl_func(pl.col("a"), pattern))
    assert_matches_polars(q)

File: cpp/tests/test_strings.cpp
/*
 * Copyright (c) 2023-2025, NVIDIA CORPORATION.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "gmock/gmock-matchers.h"
#include <arrow/api.h>
#include <arrow/compute/api.h>
#include <gtest/gtest.h>
#include <legate.h>

#include "test_utils.hpp"
#include <legate_dataframe/core/column.hpp>
#include <legate_dataframe/strings.hpp>

using namespace legate::dataframe;

void CompareArrow(const std::string& match_func,
                  const LogicalColumn& column,
                  const std::string& pattern)
{
  std::vector<arrow::Datum> args = {column.get_arrow()};
  auto options                   = arrow::compute::MatchSubstringOptions(pattern);
  auto expected =
    (ARROW_RESULT(arrow::compute::CallFunction(match_func, args, &options))).make_array();
  auto result = strings::match(match_func, column, pattern).get_arrow();

  EXPECT_TRUE(expected->Equals(*result))
    << "Failed " << match_func << ": " << column.get_arrow()->ToString() << " pattern: " << pattern
    << " Expected: " << expected->ToString() << " Result: " << result->ToString();
}

TEST(StringsTest, StartsWithBasic)
{
  LogicalColumn values(std::vector<std::string>{"hello", "world", "help", "spam"});
  CompareArrow("starts_with", values, "hel");  // codespell:ignore
}

TEST(StringsTest, EndsWithBasic)
{
  LogicalColumn values(std::vector<std::string>{"hello", "", "help", "fello"});
  CompareArrow("ends_with", values, "llo");  // codespell:ignore
}

TEST(StringsTest, Contains)
{
  LogicalColumn values(std::vector<std::string>{"hord", "world", "help"});
  CompareArrow("match_substring", values, "or");
}

TEST(StringsTest, ContainsRegex)
{
  LogicalColumn values(std::vector<std::string>{"hord", "world", "help"});
  CompareArrow("match_substring_regex", values, "o.+d");
}

TEST(StringsTest, MatchesWithNulls)
{
  LogicalColumn values(std::vector<std::string>{"lolo", "lolo", "spam", "spam"},
                       std::vector<bool>{true, false, true, false});
  CompareArrow("starts_with", values, "lo");
  CompareArrow("ends_with", values, "lo");
  CompareArrow("match_substring", values, "lo");
  CompareArrow("match_substring_regex", values, "lo");
}

TEST(StringsTest, ErrorNonStringColumn)
{
  LogicalColumn values(std::vector<int>{1, 2, 3});

  EXPECT_THROW(strings::match("starts_with", values, "he"), std::invalid_argument);
}
